### PR TITLE
Optimize concatenate

### PR DESCRIPTION
**Pull request by unnonouno**
I made an optimized kernel for ``concatenate`` method.
related to #1492

### CODE REVIEW

**Comment by unnonouno** [comment]
This is result of @jekbradbury 's script.

```
CPU split_axis ... finished in 5.35ms
CPU concat ... finished in 3.31ms
copy to GPU ... finished in 1.42ms
GPU split_axis ... finished in 7.44ms
GPU concat ... finished in 4.36ms
copy to CPU ... finished in 4.44ms
```

**Comment by unnonouno** [comment]
I also optimize the patch with binary search to find index for each thread, but it cannot improve the result.

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -1940,17 +1940,68 @@ cpdef ndarray _repeat(ndarray a, repeats, axis=None):
...
-    i = 0
+    all_same_type = True
     for a in tup:
-        aw = a.shape[axis]
-        ret[skip + (slice(i, i + aw),)] = a
-        i += aw
+        all_same_type &= a.dtype == tup[0].dtype
+
+    if all_same_type:
+        base = numpy.prod(shape[axis+1:], dtype='i')
```
Comment (okuta) [inline cupy/core/core.pyx]: `numpy.prod` is too slow. Please use  `internal.prod`.

File: cupy/core/core.pyx [hunk 2]
```diff
@@ -1940,17 +1940,68 @@ cpdef ndarray _repeat(ndarray a, repeats, axis=None):
...
+        x = array([a.data.ptr for a in tup])
```
Comment (okuta) [inline cupy/core/core.pyx]: Please use stream.
Current code has synchronize mode copy (host to device).
Thus this code is slow.

**Comment by unnonouno** [comment]
I updated both CUDA code and CPU code. Currently CPU is bottle neck.
```
CPU split_axis ... finished in 5.63ms
CPU concat ... finished in 3.30ms
copy to GPU ... finished in 1.48ms
GPU split_axis ... finished in 7.32ms
GPU concat ... finished in 3.08ms
copy to CPU ... finished in 1.21ms
```

**Comment by unnonouno** [comment]
I got this result with more optimized version of concatenate for c-contiguous arrays. We need to optimizer python code in cupy.concatenate with cython to improve it more.
```
CPU split_axis ... finished in 5.13ms
CPU concat ... finished in 3.38ms
copy to GPU ... finished in 1.41ms
GPU split_axis ... finished in 7.61ms
GPU concat ... finished in 2.48ms
copy to CPU ... finished in 1.32ms
```

File: cupy/manipulation/join.py [hunk 3]
```diff
@@ -60,22 +62,25 @@ def concatenate(tup, axis=0):
...
+        have_same_types &= (a.dtype == dtype)
         if a.ndim != ndim:
             raise ValueError(
                 'All arrays to concatenate must have the same ndim')
-        if any(i != axis and shape[i] != a.shape[i]
-               for i in six.moves.range(ndim)):
-            raise ValueError(
-                'All arrays must have same shape except the axis to '
-                'concatenate')
+        for i in six.moves.range(ndim):
```
Comment (okuta) [inline cupy/manipulation/join.py]: How about using `list`?
```
ashape = list(a.shape)
ashape[axis] = shape[axis]
if ashape != shape:
   raise...
```

File: cupy/core/core.pyx [hunk 4]
```diff
@@ -1984,18 +1984,120 @@ cpdef ndarray _repeat(ndarray a, repeats, axis=None):
...
-    skip = (slice(None),) * axis
-    i = 0
+    all_same_type = True
+    all_one_and_contiguous = True
+    dtype = tup[0].dtype
     for a in tup:
-        aw = a.shape[axis]
-        ret[skip + (slice(i, i + aw),)] = a
-        i += aw
+        all_same_type &= a.dtype == dtype
```
Comment (okuta) [inline cupy/core/core.pyx]: How about using `all_same_type = all_same_type or a.dtype == dtype`?
Using short-circuit evaluation is better.

File: cupy/core/core.pyx [hunk 5]
```diff
@@ -1984,18 +1984,120 @@ cpdef ndarray _repeat(ndarray a, repeats, axis=None):
...
+        all_one_and_contiguous &= (a._c_contiguous and a._shape[axis] == 1)
```
Comment (okuta) [inline cupy/core/core.pyx]: How about using `all_one_and_contiguous = all_one_and_contiguous or (a._c_contiguous and a._shape[axis] == 1)`?

File: cupy/core/core.pyx [hunk 6]
```diff
@@ -1984,18 +1984,120 @@ cpdef ndarray _repeat(ndarray a, repeats, axis=None):
...
+                cum += a._shape[axis]
+            cum_sizes = array(cum_sizes)
+
+            _concatenate_kernel(
+                x, axis, cum_sizes, x_strides, array(shape, 'i'), base, ret)
+    else:
+        skip = (slice(None),) * axis
+        i = 0
+        for a in tup:
+            aw = a.shape[axis]
```
Comment (okuta) [inline cupy/core/core.pyx]: Please use `a._shape`.

File: cupy/manipulation/join.py (hunk 3 above)
Comment (unnonouno) [inline cupy/manipulation/join.py]: I'll fix it in other PR. I mean I'll fix it in Cython world.

**Comment by okuta** [comment]
LGTM!

**Pull request merged**